File: airbyte-integrations/connectors/source-mssql/src/test/kotlin/io/airbyte/integrations/source/mssql/legacy/CdcMssqlSourceDatatypeTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.source.mssql.legacy;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.cdk.db.Database;
import io.airbyte.integrations.source.mssql.MsSQLTestDatabase;
import io.airbyte.integrations.source.mssql.MsSQLTestDatabase.BaseImage;
import io.airbyte.integrations.source.mssql.MsSQLTestDatabase.ContainerModifier;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;

@TestInstance(Lifecycle.PER_METHOD)
@Execution(ExecutionMode.CONCURRENT)
public class CdcMssqlSourceDatatypeTest extends AbstractMssqlSourceDatatypeTest {

  private final ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());

  @Override
  protected JsonNode getConfig() {
    return testdb.integrationTestConfigBuilder()
        .withCdcReplication()
        .withoutSsl()
        .build();
  }

  @Override
  protected Database setupDatabase() {
    testdb = MsSQLTestDatabase.in(BaseImage.MSSQL_2022, ContainerModifier.AGENT)
        .withCdc();
    return testdb.getDatabase();
  }

  protected void createTables() throws Exception {
    List<Callable<MsSQLTestDatabase>> createTableTasks = new ArrayList<>();
    List<Callable<MsSQLTestDatabase>> enableCdcForTableTasks = new ArrayList<>();
    for (var test : testDataHolders) {
      createTableTasks.add(() -> testdb.with(test.getCreateSqlQuery()));
      enableCdcForTableTasks.add(() -> testdb.withCdcForTable(test.getNameSpace(), test.getNameWithTestPrefix(), null));
    }
    executor.invokeAll(createTableTasks);
    executor.invokeAll(enableCdcForTableTasks);
  }

  protected void populateTables() throws Exception {
    List<Callable<MsSQLTestDatabase>> insertTasks = new ArrayList<>();
    List<Callable<MsSQLTestDatabase>> waitForCdcRecordsTasks = new ArrayList<>();
    for (var test : testDataHolders) {
      insertTasks.add(() -> {
        this.database.query((ctx) -> {
          List<String> sql = test.getInsertSqlQueries();
          Objects.requireNonNull(ctx);
          sql.forEach(ctx::fetch);
          return null;
        });
        return null;
      });
      waitForCdcRecordsTasks.add(() -> testdb.waitForCdcRecords(test.getNameSpace(), test.getNameWithTestPrefix(), test.getExpectedValues().size()));
    }
    // executor.invokeAll(insertTasks);
    executor.invokeAll(insertTasks);
    executor.invokeAll(waitForCdcRecordsTasks);
  }

  @Override
  public boolean testCatalog() {
    return true;
  }

}

File: airbyte-integrations/connectors/source-mssql/src/test/kotlin/io/airbyte/integrations/source/mssql/legacy/CdcMssqlSourceDatatypeTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.source.mssql.legacy

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.test.fixtures.legacy.Database
import io.airbyte.integrations.source.mssql.MsSQLTestDatabase
import io.airbyte.integrations.source.mssql.MsSQLTestDatabase.Companion.`in`
import org.jooq.DSLContext
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.parallel.Execution
import org.junit.jupiter.api.parallel.ExecutionMode
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.function.Consumer

@TestInstance(TestInstance.Lifecycle.PER_METHOD)
@Execution(ExecutionMode.CONCURRENT)
class CdcMssqlSourceDatatypeTest : AbstractMssqlSourceDatatypeTest() {
    private val executor: ExecutorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors())
    override val isCdcTest = false

    override val config: JsonNode
        get() = testdb!!.integrationTestConfigBuilder()
            .withCdcReplication()
            .withoutSsl()
            .build()

    override fun setupDatabase(): Database? {
        testdb = `in`(MsSQLTestDatabase.BaseImage.MSSQL_2022, MsSQLTestDatabase.ContainerModifier.AGENT)
            .withCdc()
        return testdb!!.database
    }

    @Throws(Exception::class)
    override fun createTables() {
        val createTableTasks: MutableList<Callable<MsSQLTestDatabase>> = ArrayList()
        val enableCdcForTableTasks: MutableList<Callable<MsSQLTestDatabase>> = ArrayList()
        for (test in testDataHolders) {
            createTableTasks.add(Callable { testdb!!.with(test.createSqlQuery) })
            enableCdcForTableTasks.add(Callable { testdb!!.withCdcForTable(test.nameSpace, test.nameWithTestPrefix, null) })
        }
        executor.invokeAll(createTableTasks)
        executor.invokeAll(enableCdcForTableTasks)
    }

    @Throws(Exception::class)
    override fun populateTables() {
        val insertTasks: MutableList<Callable<MsSQLTestDatabase?>> = ArrayList()
        val waitForCdcRecordsTasks: MutableList<Callable<MsSQLTestDatabase?>> = ArrayList()
        for (test in testDataHolders) {
            insertTasks.add(Callable {
                database!!.query<Any?> { ctx: DSLContext ->
                    val sql = test.insertSqlQueries
                    Objects.requireNonNull(ctx)
                    sql.forEach(Consumer { sql: String? -> ctx.fetch(sql) })
                    null
                }
                null
            })
            waitForCdcRecordsTasks.add(Callable { testdb!!.waitForCdcRecords(test.nameSpace, test.nameWithTestPrefix, test.expectedValues.size) })
        }
        // executor.invokeAll(insertTasks);
        executor.invokeAll(insertTasks)
        executor.invokeAll(waitForCdcRecordsTasks)
    }

    public override fun testCatalog(): Boolean {
        return true
    }
}
